### PR TITLE
Introduce an auto deploy feature

### LINKED ISSUE #20: Proposal of auto deploy
As @amiq11 suggested in [the mailing list](https://groups.google.com/forum/?hl=ja#!topic/pile-dev/yuP2n1bIX1s), I propose to use CI for auto deploy.

> 1. We should do the following 3 steps (testing/packaging/launching) AUTOMATICALLY
> 2. The triggers of them are different
>   a. testing occurs after pushing any change [DONE]
>   b. packaging occurs after merging (develop -> master)
>   c. launching occurs after making a release branch

As for this branch, `launching` means making a release and 2.b and 2.c can be done at once?

### PR DESCRIPTION
**Pull request by tiwanari**
resolves #20.

This PR introduces a feature of auto deploy.

### Procedure
1. Merge changes into `master` (e.g., `git checkout master; git merge develop`)
1. Edit the version of the project in `build.gradle` ([this line](https://github.com/PileProject/drivecommand/blob/37a4b6f7f83d76a6de077aedb7052be701df719f/build.gradle#L63))
1. Run a task (`./gradlew uploadArchives`)
1. Make a commit (e.g., `git add .; git commit -m "Release v2.2.0"`)
1. Make a tag (e.g., `git tag v2.2.0`)
1. Push changes (`git push --follow-tags`)

Then, `gh-pages` branch will be updated automatically :tada:

### References
- [promises-guide/deploy-gh-pages.sh at master · w3ctag/promises-guide]( https://goo.gl/6816am )
- (Japanese) [Travis CI経由でgh-pagesにmiddlemanで構築した静的サイトをデプロイ。 - Qiita]( https://goo.gl/GxNSIq )

### CODE REVIEW

**Comment by mandaiy** [comment]
I don't know this magical things but seems ok.
Did you test this?

**Comment by tiwanari** [comment]
Yes, I tested the procedure with my forked repository and it worked. But, strictly speaking, I haven't tested this PR because the `secure` was newly created for this repository. I hope it works properly 🙏 

#### TL; DR
I copied the script and the changes on `.travis.yml` for this PR, and made a new `personal access token` on my https://github.com/settings/tokens page by checking only `Access public repositories`.

After getting the token, I ran the following command for this repository:
```
travis encrypt -r PileProject/drivecommand "GH_TOKEN=TOKEN" --add
```
and this added some lines into `.travis.yml` automatically. The result is this PR.

**Comment by tiwanari** [comment]
By the way, I'm not sure whether `master` branch should have [release directory](https://github.com/PileProject/drivecommand/tree/master/com/pileproject/drivecommand). And, I found it now diversed from `gh-pages` branch's [release directory](https://github.com/PileProject/drivecommand/tree/gh-pages/com/pileproject/drivecommand). What do you think? :)

**Comment by tiwanari** [comment]
Ah, I remembered that the uploadArchives task refers to the existing releases by checking [maven metadata file](https://github.com/PileProject/drivecommand/blob/master/com/pileproject/drivecommand/maven-metadata.xml). So, at least we should keep the file in master and keep (restore) the consistence.

**Comment by tiwanari** [comment]
Hey @makotoshimazu @myusak 

Please give me comments or approvals:) I want to merge this and make `v2.2.0` before merging #23 and making `v2.3.0` tag for that.

File: scripts/deploy-gh-pages.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh:1]: I think we should remove these comments to avoid needless copyright violation and rewrite our own words.

File: scripts/deploy-gh-pages.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+set -e # Exit with nonzero exit code if anything fails
+
+mkdir tmp
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: Maybe it's safer way to use `mktemp` than `mkdir`, though `mktemp` is not in posix but I'm sure that the running OS is Ubuntu.

File: scripts/deploy-gh-pages.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+cd tmp
+
+SOURCE_BRANCH="master"
+TARGET_BRANCH="gh-pages"
+
+# Pull requests and commits to other branches shouldn't try to deploy, just build to verify
+if [ "$TRAVIS_PULL_REQUEST" != "false" -o -z "$TRAVIS_TAG" ]; then
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: I'd split this `if` statement and add messages why we are skipping deploy, like `echo 'Skipping deploy because $TRAVIS_TAG is missing'`
I don't know where you define these environmental variables so if my comment is pointless please ignore.

File: scripts/deploy-gh-pages.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+fi
+
+# Save some useful information
+REPO=`git config remote.origin.url`
+REPO_WITH_TOKEN=${REPO/https:\/\/github.com\//https://${GH_TOKEN}@github.com/}
+
+# Clone the existing gh-pages for this repo into out/
+# Create a new empty branch if gh-pages doesn't exist yet (should only happen on first deply)
+git clone $REPO src
+cd src; git checkout $SOURCE_BRANCH
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: I don't feel the necessity of `;`, why not remove this and write two separate commands?

File: scripts/deploy-gh-pages.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+# Build
+./gradlew uploadArchives
+
+cd ..
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: I think we should avoid `cd`'ing with relative paths.
Keep the root path of this temporary directory and use it for `cd`

File: scripts/deploy-gh-pages.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+# Copy output files
+cp -r src/com/pileproject/drivecommand/* tgt/com/pileproject/drivecommand
+
+# Now let's go have some fun with the cloned repo
+cd tgt
+git config --global user.email "travis@travis-ci.org"
+git config --global user.name "Travis"
+
+# If there are no changes (e.g. this is a README update) then just bail.
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: Maybe we can check this earlier, just below the `cd tgt`

File: scripts/deploy-gh-pages.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+if [ -z `git diff --exit-code` ]; then
+    echo "No changes to the spec on this push; exiting."
+    exit 0
+fi
+
+# Commit the "changes", i.e. the new version.
+# The delta will show diffs between new and old versions.
+git add .
+git commit -m "Release ${TRAVIS_TAG}"
+git push --quiet $REPO_WITH_TOKEN $TARGET_BRANCH
```
Comment (mandaiy) [inline scripts/deploy-gh-pages.sh]: Better to echo a message to tell we've done the deploy successfully?

**Comment by mandaiy** [comment]
I had a detailed look on your changes and now I have a few questions and comments.

Questions:
- How can we specify `$TRAVIS_TAG`?
Is it already defined in Travis?

Comments:
- I investigated `build.gradle` and I found a way to specify `uploadArchives.repositories.mavenDeployers.pom.version` via command line arguments.
If you have time please change your script (L26) and then we don't have to change `build.gradle` per our release.
With my diff, the artifacts can be generated by `./gradlew deploy -Ptag=$TAG`
```diff
diff --git a/build.gradle b/build.gradle
index a2bc4b0..c8b36ab 100644
--- a/build.gradle
+++ b/build.gradle
@@ -44,10 +44,6 @@ dependencies {
 
 }
 
-task wraper(type: Wrapper) {
-    gradleVersion = '2.5'
-}
-
 // for our license
 license {
     header rootProject.file('LICENSE')
@@ -55,26 +51,37 @@ license {
 }
 
 // Maven repository identity setting
-uploadArchives {
-    repositories {
-        mavenDeployer {
-            repository url: "file:${projectDir}"
-
-            pom.version = '2.1.0'
-            pom.groupId = 'com.pileproject'
-            pom.artifactId = 'drivecommand'
+uploadArchives.repositories.mavenDeployer {
+    repository url: "file:${projectDir}"
+    pom {
+        version = 'unspecified'
+        groupId = 'com.pileproject'
+        artifactId = 'drivecommand'
 
-            pom.project {
-                inceptionYear '2011'
-                packaging 'jar'
-                licenses {
-                    license {
-                        name 'The Apache Software License, Version 2.0'
-                        url 'http://www.apache.org/licenses/LICENSE-2.0.txt'
-                        distribution 'repo'
-                    }
+        project {
+            inceptionYear '2011'
+            packaging 'jar'
+            licenses {
+                license {
+                    name 'The Apache Software License, Version 2.0'
+                    url 'http://www.apache.org/licenses/LICENSE-2.0.txt'
+                    distribution 'repo'
                 }
             }
         }
     }
 }
+
+task deploy {
+    description = "Deploys current sourceset into .jar file via 'uploadArchives' task."
+
+    doLast {
+        if (!project.hasProperty('tag')) {
+            throw new GradleException('Tag not specified. Run with -Ptag=$TAG')
+        }
+
+        uploadArchives.repositories.mavenDeployer.pom.version = project.tag
+
+        uploadArchives.execute()
+    }
+}
```

- Now we can publish GitHub pages via `/docs` directory in `master` branch
https://help.github.com/articles/configuring-a-publishing-source-for-github-pages/
So your script could be simplified because we don't have to deal with chaining branches.

**Review by mandaiy — CHANGES_REQUESTED** [review]
Please check my comments in your script.
As I wrote in conversation, if you have time please consider enhancing your script.

File: scripts/deploy-gh-pages.sh (hunk 2 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: Thanks for the information:)

File: scripts/deploy-gh-pages.sh (hunk 1 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh:1]: 👍

File: scripts/deploy-gh-pages.sh (hunk 3 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: The list of variables is [here](https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables) and you can see "`TRAVIS_TAG: If the current build is for a git tag, this variable is set to the tag’s name.`" So, I found that I can remove this compound because I specify the trigger [here](https://github.com/PileProject/drivecommand/pull/22/files#diff-354f30a63fb0907d4ad57269548329e3R24) and that requires a tag (and thus, it cannot be a PR).

File: scripts/deploy-gh-pages.sh (hunk 4 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: Okay:)

File: scripts/deploy-gh-pages.sh (hunk 5 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: 👍

File: scripts/deploy-gh-pages.sh (hunk 6 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: I'm not sure but stuff like `git config foo` should be done before `git diff`(?) but anyway I will refactor this file.

File: scripts/deploy-gh-pages.sh (hunk 7 above)
Comment (tiwanari) [reply scripts/deploy-gh-pages.sh]: 👍

**Review by tiwanari — COMMENTED** [review]
Thank you for your review, @myusak !
They are really helpful 👍 

I will fix them and refactor scripts. Please give me a little time:)

**Comment by tiwanari** [comment]
Hi @myusak 

I fixed scripts and tested them:)

### Update
Thanks to your valuable comments, the procedure becomes much simpler 🎉 

1. Merge changes into `master` (e.g., `git checkout master; git merge develop`)
~1. Edit the version of the project in `build.gradle` ([this line](https://github.com/PileProject/drivecommand/blob/37a4b6f7f83d76a6de077aedb7052be701df719f/build.gradle#L63))~
~1. Run a task (`./gradlew uploadArchives`)~
~1. Make a commit (e.g., `git add .; git commit -m "Release v2.2.0"`)~
1. Make a tag (e.g., `git tag v2.2.0`)
1. Push changes (`git push --follow-tags`)

### Note
I specified the format of a tag for deploy to `^v[0-9].*$` and added a feature removing the prefix `v` in the task you suggested to follow the current naming rule of release folders.

### Verification
[This tag commit](https://travis-ci.org/tiwanari/drivecommand#L299) successfully deployed the release to `/docs` and [this commit](https://github.com/tiwanari/drive/commit/f17be04a61c8713e51d6c0da56e83e367da8fedb) properly changed the source of the library in [CI](https://travis-ci.org/tiwanari/drive#L2340) (and it has errors because the `v.2.2.0` changed the signature of a function).

File: .travis.yml [hunk 8]
```diff
@@ -1,18 +1,25 @@
...
-  - ./gradlew license check --continue --stacktrace
+- "./gradlew license check --continue --stacktrace"
+deploy:
+  provider: script
+  script: scripts/deploy-gh-pages.sh
+  skip_cleanup: true
+  on:
+    branch: master
+    tags: true
+    condition: "$TRAVIS_TAG =~ ^v[0-9].*$"
```
Comment (tiwanari) [inline .travis.yml]: Check the format here.

File: build.gradle [hunk 9]
```diff
@@ -44,37 +44,49 @@ dependencies {
...
+
+task deploy {
+    description = "Deploys current sourceset into .jar file via 'uploadArchives' task."
+
+    doLast {
+        if (!project.hasProperty('tag')) {
+            throw new GradleException('Tag not specified. Run with -Ptag=$TAG')
+        }
+
+        if (!(project.tag =~ /^v\d.*$/)) {
```
Comment (tiwanari) [inline build.gradle:83]: This is the same check of `.travis.yaml`.

File: .travis.yml (hunk 8 above)
Comment (tiwanari) [inline .travis.yml:25]: Check the format here.

File: build.gradle [hunk 10]
```diff
@@ -44,37 +44,49 @@ dependencies {
...
+            throw new GradleException('Wrong format tag. A tag for deploy should be /^v[0-9].*$/')
+        }
+
+        // remove the prefix 'v'
+        uploadArchives.repositories.mavenDeployer.pom.version = project.tag.substring(1)
```
Comment (tiwanari) [inline build.gradle:88]: remove `v`

**Review by tiwanari — COMMENTED** [review]
Added a few comments. 

Please review this again, @myusak

**Review by mandaiy — APPROVED** [review]
Looks good. Good job 👍

**Comment by tiwanari** [comment]
Thanks, @myusak !! :)

Let me merge it 🎉
